New: `no-global-assign` rule (fixes #6586)

alberto opened the pull request:
<!--
Thanks for submitting a pull request to ESLint. Before continuing, please be sure you've read over our guidelines:
http://eslint.org/docs/developer-guide/contributing/pull-requests

Specifically, all pull requests containing code require an **accepted** issue (documentation-only pull requests do not require an issue). If this pull request contains code and there isn't yet an issue explaining why you're submitting this pull request, please stop and open a new issue first.

Please answer all questions below.
-->

**What issue does this pull request address?**
#6586 

**What changes did you make? (Give an overview)**
Duplicate and rename `no-native-reassign` to `no-global-assign` and mark the former as deprecated.

**Is there anything you'd like reviewers to focus on?**
Let me know if the deprecation messages are enough and if the commit message and tag are correct for this kind on change.

ilyavolodin commented on lib/rules/no-extend-native.js:
```diff
@@ -1,6 +1,7 @@
 /**
  * @fileoverview Rule to flag adding properties to native object's prototypes.
  * @author David Nelson
+ * @deprecated in ESLint v3.2.0
```
Sorry, I don't remember what did we decide. Do we just leave the old rule unchanged, or should we remove all of the code and just do `module.exports = require('./no-global-assign');`? I guess former, since we don't want new changes to the rule to be propagated to deprecated version.

alberto replied:
Yes, that's what we decided in https://github.com/eslint/eslint/issues/6586#issuecomment-232503727

mysticatea commented:
Oh, we should deprecate no-native-reassign rule, not no-extend-native rule.

mysticatea commented:
Is the deprecated rule removed from rule list page?

alberto commented:
> Oh, we should deprecate no-native-reassign rule, not no-extend-native rule.

Oops! 😅  

> Is the deprecated rule removed from rule list page?

No. I don't think it should be removed since it's still available and actually part of `eslint:recommended`.

mysticatea commented:
Looks good to me. Thank you!

nzakas commented:
We also need to figure out where to list deprecated rules and add a deprecation notice to the rule docs. Possibly, we should add this into the rule `meta` too. @ilyavolodin @pedrottimark thoughts on that?

nzakas commented on lib/rules/no-global-assign.js:
```diff
@@ -0,0 +1,83 @@
+        docs: {
+            description: "disallow assignments to native objects or read-only global variables",
+            category: "Best Practices",
+            recommended: true
```
Not actually recommended, correct?

alberto replied:
Oh, you are right

nzakas commented on lib/rules/no-native-reassign.js:
```diff
@@ -12,7 +13,7 @@
     meta: {
         docs: {
-            description: "disallow assignments to native objects or read-only global variables",
+            description: "disallow assignments to native objects or read-only global variables (deprecated)",
```
Maybe we should have `deprecated: true` instead of adding to the message?

alberto replied:
Yes, this was the minimum until we decided what to do for the rules page.

ilyavolodin commented:
Yes, I think deprecated should become another meta property. I think @mysticatea already assumed as such with his PR for `eslint:all`

platinumazure commented:
I'm :+1: on `meta.deprecated` unless that will raise some other problem I'm not aware of.

vitorbal commented:
I would be willing to help out with this issue if no one minds. I also had
in my radar generating the deprecated rules list for the rules index page,
unless you're already on that, @ilyavolodin?
On Mon, Jul 25, 2016 at 9:32 PM Kevin Partington <notifications@github.com>
wrote:

> I'm 👍 on meta.deprecated unless that will raise some other problem I'm
> not aware of.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/6746#issuecomment-235059255>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAmNdnQYXLIy0_pAHjsWUOVNaAEIFxTKks5qZQ9XgaJpZM4JTGV0>
> .
>

ilyavolodin commented:
@vitorbal No, it's all yours.

alberto commented:
@nzakas do you want me to address the rules list generation as part of this PR? Or what do you intend to do @vitorbal?

alberto commented:
I have updated the PR with the suggestions above and added a new section `deprecated` in `conf/category-list.json`

@vitorbal now that I had a look at how the generation is done, I realized that maybe you referred to the changes in the site repo?

Let me know if there is something else needed here

vitorbal commented:
@alberto I think you actually got it covered now with your latest commits.
As for the site repo, I don't have much experience there yet, but I can give it a shot! From what I can see the only thing missing would be to add the list of deprecated rules in `docs/rules/index.md`?

vitorbal commented on conf/category-list.json:
```diff
@@ -8,6 +8,13 @@
     ],
+    "deprecated": {
+        "name": "Deprecated",
+        "description": "These rules have been deprecated and replaced by newer rules.",
```
For the "removed" rules we end the description with a colon instead of a dot. Should we do the same here for consistency?

nzakas commented on Makefile.js:
```diff
@@ -209,6 +209,10 @@ function generateRuleIndexPage(basedir) {
     find(path.join(basedir, "/lib/rules/")).filter(fileType("js")).forEach(function(filename) {
         var rule = require(filename);
 
+        if (rule.meta.deprecated) {
```
Do we want to skip this entirely?  Shouldn't there be a category for that?

vitorbal replied:
We could add it to the `categoriesData.deprecated.rules` array here instead of doing it manually in https://github.com/eslint/eslint/pull/6746/files#r72645142

nzakas replied:
I think that is a good idea. Less manual work is always preferable.

nzakas commented on conf/category-list.json:
```diff
@@ -8,6 +8,13 @@
+        "name": "Deprecated",
+        "description": "These rules have been deprecated and replaced by newer rules.",
+        "rules": [
+            { "deprecated": "no-native-reassign", "replacedBy": ["no-global-assign"] }
```
Can this be calculated instead? The fewer files we have to update to deprecate something, the fewer mistakes we will make.

nzakas commented on lib/rules/no-native-reassign.js:
```diff
@@ -17,6 +18,8 @@ module.exports = {
             recommended: true
         },
 
+        deprecated: true,
```
Should this be in `docs` too? Maybe with a `replacedBy: "no-global-assign"` field?

alberto replied:
I put it there because it's what other suggested and what @mysticatea implemented in https://github.com/eslint/eslint/pull/6756. That's an example of something which is not strictly documentation. That said, I don't have a problem to move it. If we add `replaceBy`, it might make more sense to keep them together in `docs`.

I should probably keep `replacedBy` as an array, just in case a rule is replaced by many.

platinumazure replied:
Personally, I think `deprecated` goes beyond being a documentation-level property because it has extra semantic meaning outside of docs. There's no reason documentation generators have to restrict themselves to `meta.docs`, is there? Just my two cents.

ilyavolodin replied:
Agree with @platinumazure `deprecated` should be outside of docs, because it's used for other purposes then just documentation. For example we are excluding deprecated rules from `eslint:all` based on this key.

nzakas replied:
Good points, I didn't realize it was being used outside of docs. 👍

alberto commented:
PR updated with suggestions from above.

vitorbal commented:
Lgtm

nzakas commented:
LGTM
